Topics/Tech_Stacks/Unity_ML_Agents.md

```diff
@@ -0,0 +1,199 @@
+
+In order to do so, our agent will follow a simple cycle of 3 steps. It will first collect an observation from the environment. Using this information collected, it will perfmr an action. Finally, based on this action, we will give the agent a reward.
+
+Our agent will attempt to learn a strategy that will maximize the reward it receives over time.
+
+
+# Setting up the Environment
+
+In this step by step tutorial, we will learn how to create a simple environment in which we are able to train an agent to play a simple game. The basic setup I will use to introduce the concepts of reinforcement learning will involve an **Agent**, and **two rewards** of varying sizes. 
+
+<img src="https://raw.githubusercontent.com/Dismany0/misc_images/main/CSC301%20Assignment%202/Screenshot%202023-11-26%20012237.png" height="500"/>
+
+In my setup, I have currently created a floor, an Agent, and 2 rewards.
+
+They are positioned along a straight line. Our environment will involve the agent being able to perform 3 actions, moving left, moving right, or doing nothing at all. Hopefully, we will be able to train this agent to move towards the larger value.
+
+Finally, to finish this section, create a C# script and attach it to your agent object. I have named the script ```BiggerRewardAgent.cs```, but feel free to change the name to anything you'd like.
+
+# Actions
+
+Now that our environment is set up, we will assign actions that our agent is able to perform. Begin by opening the ```CollectRewardAgent.cs``` script with any text editor you would like, such as Visual Studio or Visual Studio Code. We will begin by changing our MonoBehaviour into a Agent class. Make the following imports:
+
+```
+using UnityEngine;
+using Unity.MLAgents;
+using Unity.MLAgents.Sensors;
+using Unity.MLAgents.Actuators;
+```
+
+We will also removing the ```Update()``` and ```Start()``` functions, and add the following overrides:
+
+```public override void OnEpisodeBegin()```
+
+```public override void CollectObservations(VectorSensor sensor)```
+
+```public override void OnActionReceived(ActionBuffers actions)```
+
+```public void OnTriggerEnter(Collider other)```
+
+In the following few sections, we will go through and explain each of these functions in detail. We will first begin with ```OnActionReceieved()```
+
+Unlike us, our agent is only able to play through making actions through numbers. These numbers will be given through the ActionBuffer. As specified before, we want our agent to be able to move left, right, and stand still. We will accomplish this by assigning each action a number. We begin by obtaining the action from DiscreteActions, or integer actions. We will set up where our AI gets these DiscreteActions later in our observations section.
+
+```
+public override void OnActionReceived(ActionBuffers actions)
```

Comment:
You could add a comment or two here for people to understand what's happening

Reply:
Added another small section explaining how the OnActionReceived function works